### PR TITLE
Adding Qobject tests for simulators

### PR DESCRIPTION
**Pull request by delapuente**
### Summary
This PR adds tests for simulators supporting Qobject. They compare remote execution with a local reference.

### CODE REVIEW

**Comment by delapuente** [comment]
Put on hold until the `ibmq_qasm_simulator` inform about the Qobject support correctly (now is returning that Qobject is not supported).

**Comment by taalexander** [comment]
I had to submit [a PR](https://github.com/delapuente/qiskit-terra/pull/17) against @delapuente's  branch as I did not have contributor access

**Comment by taalexander** [comment]
What I propose we do:
- `_get_credentials` returns a list of all discovered credentials. 
    - If `USE_ALTERNATE_ENV_CREDENTIALS` is set: 
          - A list of credentials is populated from the environment variable `IBMQ_CREDENTIALS`. This 
             is a string of the form `"url_1:token_1,url_2:token_2,..."`.
- Rename `requires_qe_credentials` -> `requires_ibm_experience` 
- `requires_ibm_experience`  calls `_get_credentials`
     - Converts credentials to lists of `urls` and `tokens`
     - Calls decorated function with additional `kwargs` of `urls` and `tokens`.
- `requires_qe_access` is  a decorator and its wrapper function is decorated by `requires_ibm_access`. `requires_qe_access` looks for the quantum experience credentials by its known url. If found it passes the quantum experience url and token as keyword arguments to the function it wraps. 

Then to write tests that run over multiple backends I propose using a decorator as shown below:
```
def once_per_qobj_backend(test):
    """
    Test Qobj support on all backends claiming to support Qobj.
    This way VCR creates a single cassette for each test.
    """

    @requires_qe_access
    @functools.wraps(test)
    def _wrapper(self, *args, qe_tokens=None, qe_urls=None, **kwargs):

        for qe_token, qe_url in zip(qe_tokens, qe_urls):
            IBMQ.enable_account(qe_token, qe_url)
        for backend in IBMQ.backends():
            if not backend.configuration()['allow_q_object']:
                if backend.configuration()['simulator']:
                    test = slow_test(test)
                with self.subTest(backend=backend):
                    test(self, backend, *args, **kwargs)
    return _wrapper
```
This layout should:
- Not change any existing tests that use `requires_qe_access`. 
- Allow multiple credentials to be passed to tests with minimal code duplication. 
- Allow `VCR` to track tests with multiple accounts. 
- Make writing of tests over multiple backends simple with the use of `once_per_qobj_backend`.

It will require:
- Changing the environment variables from which the credentials are read.

**Comment by delapuente** [comment]
> I had to submit a PR against @delapuente's branch as I did not have contributor access

You should have access now.

**Comment by delapuente** [comment]
For the reasons stated 

I would not worry about `VCR`. `VCR` was introduced as an experimental feature that should accomodate our use-cases, not the other way around. Talk to @1ucian0 to see how we can improve the use of `VCR` to also enable a comfy way of doing multi-backend tests.

**Comment by delapuente** [comment]
For some context and other issues with the proposed implementation, see https://github.com/delapuente/qiskit-terra/pull/17#pullrequestreview-162764005.

The `requires_qe_access` decorator uses the discovering credentials mechanism of IBMQ credentials so changing the environment variables names would effectively change the `IBMQ` provider API.

Since this is only to stress our backends and it is unlikely that regular users will ever run this test suite, I would put special effort into avoiding this test-specific change bleeds to the public API.

If you're worried about code repetition, here is a proposal that implements both decorators without repeating code:

```python
def requires_qe_access(func):

    @_online_test
    @functools.wraps(func)
    def _wrapper(self, *args, **kwargs):
        credentials = _get_credentials(self, TEST_OPTIONS)
        self.using_ibmq_credentials = credentials.is_ibmq()
        kwargs.update({'qe_token': credentials.token, 'qe_url': credentials.url})
        return func(self, *args, **kwargs)

    return _wrapper


def require_multiple_credentials(func):
    """Parses IBMQ_CREDENTIALS and passes it to the tests as a list of
    tuples (url, token)."""

    @_online_test
    @functools.wraps(func)
    def _wrapper(self, *args, **kwargs):
        credentials = os.getenv('IBMQ_CREDENTIALS', '')
        credentials = [
            entry.split(':') for entry in credentials.split(',')] if credentials else []
        kwargs.update({'credentials': credentials})
        return func(self, *args, **kwargs)

    return _wrapper


def _online_test(func):

    @functools.wraps(func)
    def _wrapper(self):
        if TEST_OPTIONS['skip_online']:
            raise unittest.SkipTest('Skipping online tests')

        decorated_func = func
        if TEST_OPTIONS['rec'] or TEST_OPTIONS['mock_online']:
            # For recording or for replaying existing cassettes, the test should be decorated with
            # use_cassette.
            decorated_func = VCR.use_cassette()(decorated_func)

        return decorated_func(self)

    return _wrapper
```

I would not worry about `VCR`. `VCR` was introduced as an experimental feature that should accomodate our use-cases, not the other way around. Talk to @1ucian0 to see how we can improve the use of `VCR` to also enable a comfy way of doing multi-backend tests.

Perhaps the testing environment requires a broader rethink and this can be part of it but without previous discussion, it is hard to know what we need and I'll try not to introduce a lot of changes with this PR. As explained in the review, this use-case, in particular, seems very specific to me to justify a change in our testing infrastructure.

I would ask for a second opinion here, @diego-plan9 do you mind to take a look and share your thoughts about the implementation?

**Comment by diego-plan9** [comment]
Without being fully up to date with the inner-most details, looking at the issue from a general point of view  I also think @delapuente's last comment proposes a more sensible trade-off than the proposal at https://github.com/delapuente/qiskit-terra/pull/17. 

I think it boils down to weighing how "central" the tests on this PR are compared to the rest of the test suite, and if I'm reading it correctly, they do seem to conform to a very specific use case (specially if they are to be marked as slow tests and mostly used as a mechanism for "manually" controlling the adoption of qobj by the backends) and maybe short-lived depending on the strategy (ie. ideally, we would reach a stage where all backends support qobj, so the rest of the tests take over).

If that is the case, I think updating the general test "api" ("api" might be too big a word here) and as a result transferring the complexity to the rest of the test writers and maintainers seems too taxing - I would indeed recommend being conservative: treat them as a special case via a custom decorator trying to minimize the impact on the rest of the tests, and if eventually we see the need for evolving the test suite, reevaluate the decision. Would that make sense? Actually pinging @ajavadia as he has been more involved in this particular issue.

**Comment by ajavadia** [comment]
Let me clarify the purpose of these tests a bit. They are meant for checking the functionality of the experiments. They will continue to exist in some form or another (eventually in the IBMQProvider standalone repo). The experiment-side codes continually change (in the near term for supporting qobj, and for other features in the long run). We need integration tests that exercise different circuits on different IBMQ devices.

So I modify my original statement. This is more than just testing qobj. We need tests that run on every device once (but don't duplicate if a single device is accessible through multiple accounts). Currently we do not have such a test. I would actually put it in `test/python/ibmq/test_ibmq_devices.py`.

Since these tests will soon live in another repo, I agree with a minimally invasive approach that doesn't change Terra's tests much. If a custom decorator achieves this, then great.

**Comment by 1ucian0** [comment]
Just a friendly ping. Is this still a WIP?

File: test/python/ibmq/test_ibmq_qobj.py [hunk 1]
```diff
@@ -13,88 +13,87 @@
...
 from qiskit import (ClassicalRegister, QuantumCircuit, QuantumRegister, compile)
-
 from qiskit import IBMQ, Aer
 from qiskit.qasm import pi
-from ..common import requires_qe_access, JobTestCase, slow_test
+
+from ..common import require_multiple_credentials, JobTestCase, slow_test
+
+# Timeout duration
+TIMEOUT = os.getenv("IBMQ_TIMEOUT", 10)
```
Comment (delapuente) [inline test/python/ibmq/test_ibmq_qobj.py]: Is `IBMQ_TIMEOUT` environment variable used elsewhere? If not and if it is specific for tests, can we change the name to `IBMQ_TESTS_TIMEOUT` or something more explicit?

File: test/python/ibmq/test_ibmq_qobj.py [hunk 2]
```diff
@@ -13,88 +13,87 @@
...
+    @functools.wraps(test)
+    def _wrapper(self, *args, credentials=[], **kwargs):
+        for qe_token, qe_url in credentials:
+            IBMQ.enable_account(qe_token, qe_url)
+        for backend in IBMQ.backends():
+            config = backend.configuration()
+            if config['allow_q_object']:
+                with self.subTest(backend=backend):
+                    backend_test = test if config['simulator'] else slow_test(test)
+                    backend_test(self, backend, *args, **kwargs)
```
Comment (delapuente) [inline test/python/ibmq/test_ibmq_qobj.py]: If possible, I think it's better if we disable the accounts after the test?

File: test/python/ibmq/test_ibmq_qobj.py (hunk 1 above)
Comment (taalexander) [reply test/python/ibmq/test_ibmq_qobj.py]: I will change it to this.

File: test/python/common.py [hunk 3]
```diff
@@ -15,6 +15,7 @@
 import time
 import unittest
 from unittest.util import safe_repr
+import regex as re
```
[the remark this answers is not in the document]
Comment (taalexander) [reply test/python/common.py]: Sorry, just muscle memory typing that.

File: test/python/ibmq/test_ibmq_qobj.py (hunk 2 above)
Comment (taalexander) [reply test/python/ibmq/test_ibmq_qobj.py]: Agreed, won't hurt.

File: test/python/ibmq/test_ibmq_qobj.py [hunk 4]
```diff
@@ -105,17 +104,19 @@ def test_simple_circuit(self):
...
-        """Test one circuit, one register, out-of-order readout.
-        """
+                                   result_local.get_counts(circ), delta=100)
+
+    @once_per_qobj_backend
+    def test_readout_order(self, remote_backend):
+        """Test one circuit, one register, out-of-order readout."""
+        config = remote_backend.configuration()
+        n_qubits = config['n_qubits']
+        if n_qubits < 4 or config.get('n_registers', n_qubits) < 4:
```
[the remark this answers is not in the document]
Comment (taalexander) [reply test/python/ibmq/test_ibmq_qobj.py]: I don't think to so, certain tests will require specific backend features, ie. two-qubit gates, multiple registers, conditionals.

File: test/python/ibmq/test_ibmq_qobj.py [hunk 5]
```diff
@@ -192,50 +199,54 @@ def test_multi_circuit(self):
...
+                                   result_local.get_counts(circ), delta=100)
 
-    @slow_test
-    @requires_qe_access
-    def test_atlantic_circuit(self):
-        """Test Atlantis deterministic ry operation.
-        """
+
+    @once_per_qobj_backend
+    def test_atlantic_circuit(self, remote_backend):
```
[the remark this answers is not in the document]
Comment (taalexander) [reply test/python/ibmq/test_ibmq_qobj.py]: Agreed will change, I'll leave something in the docstring about why this test is special though.

File: test/python/ibmq/test_ibmq_qobj.py [hunk 6]
```diff
@@ -13,88 +13,99 @@
...
-
 from qiskit import IBMQ, Aer
 from qiskit.qasm import pi
-from ..common import requires_qe_access, JobTestCase, slow_test
+
+from ..common import require_multiple_credentials, JobTestCase, slow_test
+
+# Timeout duration
+TIMEOUT = int(os.getenv("IBMQ_TESTS_TIMEOUT", 10))
+BLACKLIST = [bl.strip() for bl in
```
Comment (delapuente) [inline test/python/ibmq/test_ibmq_qobj.py]: We use environment variables to keep credentials secret but I think there is no need of keeping the blacklist secret. Can we make the variable an explicit list defined here?

File: qiskit/backends/ibmq/api/ibmqconnector.py [hunk 7]
```diff
@@ -590,7 +590,12 @@ def get_job(self, id_job, hub=None, group=None, project=None,
 
         job = self.req.get(url)
 
-        if 'qasms' in job:
+        if 'qObjectResult' in job:
```
Comment (delapuente) [inline qiskit/backends/ibmq/api/ibmqconnector.py]: Has something changed in the backend? Why is this change necessary?

File: test/python/ibmq/test_ibmq_qobj.py (hunk 6 above)
Comment (taalexander) [reply test/python/ibmq/test_ibmq_qobj.py]: Ok, the same for TIMEOUT?

File: qiskit/backends/ibmq/api/ibmqconnector.py (hunk 7 above)
Comment (taalexander) [reply qiskit/backends/ibmq/api/ibmqconnector.py]: This is a cherrypick of #1263.

**Comment by delapuente** [comment]
@taalexander please, do you mind to fix the style problem?

>test/python/ibmq/test_ibmq_qobj.py:61:1: E302 expected 2 blank lines, found 1
make[3]: *** [style] Error 1
make[2]: *** [CMakeFiles/style.dir/all] Error 2
make[1]: *** [CMakeFiles/style.dir/rule] Error 2
make: *** [style] Error 2
The command "make style VERBOSE="" && make lint VERBOSE=""" exited with 2.

**Comment by taalexander** [comment]
This is blocking on #1279

**Comment by delapuente** [comment]
This looks great to me but I can not approve the PR because I open the PR. Can someone else do it for me once #1279 is done? @diego-plan9, @atilag?

File: test/python/ibmq/test_ibmq_qobj.py [hunk 8]
```diff
@@ -13,88 +13,101 @@
...
+                        backend_test(self, backend, *args, **kwargs)
+            for qe_token, _ in credentials:
+                IBMQ.disable_accounts(token=qe_token)
+        return _wrapper
+    return per_qobj_backend_decorator
+
+
+# no blacklist decorator.
+PER_BACKEND = per_non_blacklisted_backend()
+PER_BACKEND_BLACKLISTED = per_non_blacklisted_backend(*BLACKLIST)
```
Comment (delapuente) [inline test/python/ibmq/test_ibmq_qobj.py]: This name is misleading (it seems to mean "for every blacklisted backend"). What about using functions to avoid the capitalized names and have two decorators `@per_qobj_backend` and `@per_restricted_qobj_backend`.

**Comment by delapuente** [comment]
@ajavadia I fixed the names myself. Would you mind to try the branch locally, with a list of valid credentials, and grant the r+ if everything is OK? I don't have any further comments.

**Comment by ajavadia** [comment]
I will revisit this after #1279 where we can process hex results.

**Comment by ajavadia** [comment]
Removing the on hold label. Once @taalexander pushes his changes and confirms that everything passes, we can finally merge.

**Comment by taalexander** [comment]
Closing as we have decided to keep these tests separate from the Qiskit repository.
